packages/connect/src/protocol-grpc-web/handler-factory.spec.ts

```diff
@@ -0,0 +1,120 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { MethodInfo, ServiceType } from "@bufbuild/protobuf";
+import { Int32Value, MethodKind, StringValue } from "@bufbuild/protobuf";
+import type { MethodImpl } from "../implementation.js";
+import { createMethodImplSpec } from "../implementation.js";
+import type { UniversalHandlerOptions } from "../protocol/index.js";
+import {
+  createAsyncIterable,
+  createUniversalHandlerClient,
+  pipeTo,
+  sinkAll,
+} from "../protocol/index.js";
+import { createHandlerFactory } from "./handler-factory.js";
+import { createTransport } from "./transport.js";
+
+describe("createHandlerFactory()", function () {
+  const testService = {
+    typeName: "TestService",
+    methods: {
+      foo: {
+        name: "Foo",
+        I: Int32Value,
+        O: StringValue,
+        kind: MethodKind.Unary,
+      },
+      bar: {
+        name: "Bar",
+        I: Int32Value,
+        O: StringValue,
+        kind: MethodKind.ServerStreaming,
+      },
+    },
+  } satisfies ServiceType;
+  function setupTestHandler<M extends MethodInfo>(
+    method: M,
+    opt: Partial<UniversalHandlerOptions>,
+    impl: MethodImpl<M>
+  ) {
+    const h = createHandlerFactory(opt)(
+      createMethodImplSpec(testService, method, impl)
+    );
+    const t = createTransport({
+      httpClient: createUniversalHandlerClient([h]),
+      baseUrl: "https://example.com",
+      readMaxBytes: 0xffffff,
+      writeMaxBytes: 0xffffff,
+      compressMinBytes: 0xffffff,
+      useBinaryFormat: true,
+      interceptors: [],
+      acceptCompression: [],
+      sendCompression: null,
+    });
+    return {
+      service: testService,
+      method: method,
+      handler: h,
+      transport: t,
+    };
+  }
+
+  describe("returned handler", function () {
+    it("should surface headers for unary", async function () {
```

Comment:
I really like that after reading and understanding one test, you've read and understood them all!